codec/address/bech32_codec.go

```diff
@@ -2,30 +2,79 @@ package address
 
 import (
 	"errors"
+	"fmt"
 	"strings"
+	"sync"
+
+	"github.com/hashicorp/golang-lru/simplelru"
 
 	"cosmossdk.io/core/address"
 	errorsmod "cosmossdk.io/errors"
 
+	"github.com/cosmos/cosmos-sdk/internal/conv"
 	sdkAddress "github.com/cosmos/cosmos-sdk/types/address"
 	"github.com/cosmos/cosmos-sdk/types/bech32"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 )
 
+var errEmptyAddress = errors.New("empty address string is not allowed")
+
+type Option func(*Options)
+
+type Options struct {
+	mu    *sync.Mutex
+	cache *simplelru.LRU
+}
+
+func WithLRU(cache *simplelru.LRU) Option {
+	return func(o *Options) {
+		o.cache = cache
+	}
+}
+
+func WithMutex(mu *sync.Mutex) Option {
+	return func(o *Options) {
+		o.mu = mu
+	}
+}
+
 type Bech32Codec struct {
 	Bech32Prefix string
 }
 
-var _ address.Codec = &Bech32Codec{}
+type cachedBech32Codec struct {
+	codec Bech32Codec
+	mu    *sync.Mutex
+	cache *simplelru.LRU
+}
+
+var (
+	_ address.Codec = &Bech32Codec{}
+	_ address.Codec = &cachedBech32Codec{}
+)
+
+func NewBech32Codec(prefix string, opts ...Option) address.Codec {
+	options := Options{}
+	for _, optionFn := range opts {
+		optionFn(&options)
+	}
 
-func NewBech32Codec(prefix string) address.Codec {
-	return Bech32Codec{prefix}
+	ac := Bech32Codec{prefix}
+	if options.mu == nil || options.cache == nil {
```

Comment:
nit: I'm kind of confused why these are optional if they're both required for caching behavior.  Is a `NewCachedBech32Codec` easier to reason about?

Reply:
The idea was to have just one constructor but since there aren't any more globals having `NewBech32Codec` and `NewCachedBech32Codec` makes sense.